### ISSUE TITLE
Committee for pair potential

### ISSUE THREAD

**Issue by casv2** [ACEsuit/ACE1.jl#42]
I think it's necessary to have the committee also work for V2+ACE potentials, so include the pair potential

**Comment by casv2** [comment]
Basically what I do currently is populate some empty array using threading as follows

```
 Fs = Vector(undef, nIPs)

  @Threads.threads for i in 1:nIPs
      Fs[i] = forces(IPs[i], at)
  end
```

And then try to get some norm differences like this

```
[norm.(Fs[m] - F) for m in 1:length(Fs)]...)
```

The new threading goes over the sites, which will most likely already be much better, but in taking those force difference norms, e.g. the 'width' of the committee relative to the mean I wonder whether there's a more computationally friendly way of doing this?

**Comment by cortner** [comment]
#43  - ready to merge.

**Comment by casv2** [comment]
Maybe I'm doing something wrong, but I can't reproduce old energies from previous fitting methods. Here's a MWE

```
using ACE1
using JSON
using JuLIP

at = bulk(:Al)

Vref = OneBody(:Al => -105.8114973092, :Si => -163.2225204255)

r0 = 0.50*(rnn(:Al) + rnn(:Si))

Bsite = rpi_basis(species = [:Si, :Al],
                    N = 2,       # correlation order = body-order - 1
                    maxdeg = 13,  # polynomial degree
                    r0 = 0.3,     # estimate for NN distance
                    rin = R,
                    rcut = 5.5,   # domain for radial basis (cf documentation)
                    pin = 2)                     # require smooth inner cutoff

Bpair = pair_basis(species = [:Al, :Si],
                    r0 = r0,
                    maxdeg = 13,
                    rcut = 6.0,
                    rin = 0.0,
                    pin = 0 )   # pin = 0 means no inner cutoff

B = JuLIP.MLIPs.IPSuperBasis([Bpair, Bsite]);

X = rand(1000, length(B))
Y = rand(1000)

BRR = pyimport("sklearn.linear_model")["BayesianRidge"]

clf = BRR(normalize=true)
clf.fit(X,Y)
c = clf.coef_

IP = JuLIP.MLIPs.SumIP(Vref, JuLIP.MLIPs.combine(B, c))

old_E = energy(IP, at)

pair_pot_com = ACE1.committee_potential(Bpair, c[1:length(Bpair)], hcat([c, c, c]...)[1:length(Bpair), :])
pair_pot_com_E = ACE1.co_energy(pair_pot, at)[1]

ace_pot_com = ACE1.committee_potential(Bsite, c[length(Bpair):end], hcat([c, c, c]...)[length(Bpair):end, :])
ace_pot_com_E = ACE1.co_energy(ace_pot, at)[1]

new_E = energy(IP, at) + pair_pot_com_E + ace_pot_com_E 

new_E ≈ old_E
```

Essentially I create some fake coefficients and populate an 'old' potential with the fake c vector, then split the c vector over the pair+ACE and recalculate energies. It seems odd that any coefficient vector size seems to get accepted to form a committee? Not sure what's going on here

**Comment by cortner** [comment]
There were several bugs in your script, here is a corrected one: 
```julia
using ACE1
using JSON
using JuLIP

at = bulk(:Al)

Vref = OneBody(:Al => -105.8114973092, :Si => -163.2225204255)

r0 = 0.50*(rnn(:Al) + rnn(:Si))
rin = 0.5*r0

Bsite = rpi_basis(species = [:Si, :Al],
                    N = 2,       # correlation order = body-order - 1
                    maxdeg = 13,  # polynomial degree
                    r0 = 0.3,     # estimate for NN distance
                    rin = rin,
                    rcut = 5.5,   # domain for radial basis (cf documentation)
                    pin = 2)                     # require smooth inner cutoff

Bpair = pair_basis(species = [:Al, :Si],
                    r0 = r0,
                    maxdeg = 13,
                    rcut = 6.0,
                    rin = 0.0,
                    pin = 0 )   # pin = 0 means no inner cutoff

B = JuLIP.MLIPs.IPSuperBasis([Bpair, Bsite]);

X = rand(1000, length(B))
Y = rand(1000)


# BRR = pyimport("sklearn.linear_model")["BayesianRidge"]
# clf = BRR(normalize=true)
# clf.fit(X,Y)
# c = clf.coef_

c = randn(length(B))
c = c ./ (1:length(c)).^2

IP = JuLIP.MLIPs.SumIP(Vref, JuLIP.MLIPs.combine(B, c))

old_E = energy(IP, at)

pair_pot_com = ACE1.committee_potential(Bpair, c[1:length(Bpair)], hcat([c, c, c]...)[1:length(Bpair), :])
pair_pot_com_E = ACE1.co_energy(pair_pot_com, at)[1]

ace_pot_com = ACE1.committee_potential(Bsite, c[length(Bpair)+1:end], hcat([c, c, c]...)[length(Bpair):end, :])
ace_pot_com_E = ACE1.co_energy(ace_pot_com, at)[1]

new_E = energy(Vref, at) + pair_pot_com_E + ace_pot_com_E 

new_E ≈ old_E
```

**Comment by cortner** [comment]
what remains is the question why we can pass committee coefficients that are the wrong size. I'll look into that.

**Comment by cortner** [comment]
... this is now fixed. The script above will now fail with an error. The following line fixes it: 
 ```julia
ace_pot_com = ACE1.committee_potential(Bsite, c[length(Bpair)+1:end], hcat([c, c, c]...)[length(Bpair)+1:end, :])
```

**Comment by cortner** [comment]
Do you need a one-body with committee? I hadn't thought of this. It is a bit more awkward...

**Comment by casv2** [comment]
Thanks, I think one-body committee could be useful if people want to evaluate in Python. A bit strange if the energies don't match the mean at all...

Anyways the threading works and I'm seeing a nice factor of 5 speed up.

**Comment by cortner** [comment]
> A bit strange if those energies don't match the mean at all...

I don't understand, can you please clarify?

**Comment by casv2** [comment]
I think people will want to evaluate the committee and examine the energies from a committee potential seperately. If the `Vref` energy shift is not included in those `k` committee member energy predictions, that may cause confusion.

**Comment by casv2** [comment]
However, I'm actually happy for this to be merged as it is currently. Thanks!

**Comment by cortner** [comment]
closed by #43 and #44
